pkg/datatx/rclone/rclone.go

```diff
@@ -0,0 +1,354 @@
+}
+
+// rcloneStatusResJSON the rclone job/status method json response
+type rcloneStatusResJSON struct {
+	Finished  bool    `json:"finished"`
+	Success   bool    `json:"success"`
+	ID        int64   `json:"id"`
+	Error     string  `json:"error"`
+	Group     string  `json:"group"`
+	StartTime string  `json:"startTime"`
+	EndTime   string  `json:"endTime"`
+	Duration  float64 `json:"duration"`
+	// think we don't need this
+	// "output": {} // output of the job as would have been returned if called synchronously
+}
+
+// GetTransferStatus returns the status of the transfer with the specified job id
+func (driver *rclone) GetTransferStatus(jobID int64) (datatx.TxInfo_Status, error) {
+	rcloneStatusReq := &rcloneStatusReqJSON{
+		JobID: jobID,
+	}
+
+	data, err := json.Marshal(rcloneStatusReq)
+	if err != nil {
+		return datatx.TxInfo_STATUS_INVALID, errors.Wrap(err, "error marshalling rclone req data")
+	}
+
+	transferFileMethod := "/job/status"
+
+	u, err := url.Parse(driver.config.Endpoint)
+	if err != nil {
+		err = errors.Wrap(err, "json: error parsing driver endpoint")
+		return -1, err
+	}
+	u.Path = path.Join(u.Path, transferFileMethod)
+	requestURL := u.String()
+
+	req, err := http.NewRequest("POST", requestURL, bytes.NewReader(data))
+	if err != nil {
+		return datatx.TxInfo_STATUS_INVALID, errors.Wrap(err, "json: error framing post request")
+	}
+	req.Header.Set("Content-Type", "application/json")
+
+	req.SetBasicAuth(driver.config.AuthUser, driver.config.AuthPass)
+
+	res, err := driver.client.Do(req)
+	if err != nil {
+		err = errors.Wrap(err, "json: error sending post request")
+		return datatx.TxInfo_STATUS_INVALID, err
+	}
+
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusOK {
+		// TODO "job not found" also gives a 500
+		// Should that return STATUS_INVALID ??
+		// at the minimum the returned error message should be the rclone error message
+		resBody, e := ioutil.ReadAll(res.Body)
+		if e != nil {
+			e = errors.Wrap(e, "json: error reading response body")
+			return datatx.TxInfo_STATUS_INVALID, e
+		}
+		err = errors.Wrap(errors.New(fmt.Sprintf("%s: %s", res.Status, string(resBody))), "json: rclone request responded with error")
+		return datatx.TxInfo_STATUS_INVALID, err
+	}
+
+	var resData rcloneStatusResJSON
+	if err = json.NewDecoder(res.Body).Decode(&resData); err != nil {
+		return datatx.TxInfo_STATUS_INVALID, errors.Wrap(err, "json: error decoding response data")
+	}
+
+	if resData.Error != "" {
+		return datatx.TxInfo_STATUS_INVALID, errors.New(resData.Error)
+	}
+
+	if resData.Finished && resData.Success {
+		return datatx.TxInfo_STATUS_TRANSFER_COMPLETE, nil
+	}
+	if !resData.Finished {
+		return datatx.TxInfo_STATUS_TRANSFER_IN_PROGRESS, nil
+	}
+	if !resData.Success {
+		return datatx.TxInfo_STATUS_TRANSFER_FAILED, nil
+	}
+	return datatx.TxInfo_STATUS_INVALID, nil
+}
+
+// rcloneCancelTransferReqJSON the rclone job/stop method json request
+type rcloneCancelTransferReqJSON struct {
```

Comment:
Can we use a generic struct for rcloneCancelTransferReqJSON and rcloneStatusReqJSON, since they have the same fields? (same for rcloneCancelTransferResJSON)

Reply:
I will declare them within the function there used in, that's actually what I wanted but thought it was invalid to declare a struct inside a method (my lack of golang knowledge).